# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Daniel Zapata  <br>
Url: https://github.com/.../03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEMINARIO<br>
GitHub: https://github.com/danizagra/Algoritmos_Seminario<br>
Problema:

>3. Combinar cifras y operaciones

Descripción del problema:(copiar enunciado)


(*) La respuesta es obligatoria





                                        

## Descripción del Problema

El problema consiste en analizar y diseñar un algoritmo que resuelva la siguiente situación:

### Elementos Disponibles
- **Cifras**: Las 9 cifras del 1 al 9 (excluimos el cero)
- **Operadores**: Los 4 signos básicos de las operaciones fundamentales:
  - Suma (+)
  - Resta (-)
  - Multiplicación (*)
  - División (/)

### Objetivo
Debemos combinarlos **alternativamente** sin repetir ninguno de ellos para obtener una cantidad dada.

### Ejemplo
Para obtener el valor **4**:
4+2-6/3*1 = 4

 <span style="color: #4ECDC4; font-weight: bold;">(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?</span><br>
 
 
 
<span style="color: #4ECDC4; font-weight: bold;">¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.</span>




Respuesta

1. Teniendo en cuenta el enunciado vamos a utilizar 5 cifras y 4 operadores, que van a ser alternados entre ellos.

- **para las 5 posiciones de las cifras**: disponemos de 9 cifras distintas (1 al 9). Necesitamos elegir 5 de ellas y ordenarlas, por lo tanto tenemos una permutación de 9 elementos tomados de 5 en 5. 
    $$
    P(9,5) = \frac{9!}{(9 - 5)!} = \frac{9!}{4!} = 9 x 8 x 7 x 6 x 5 = 15120 posibilidades
    $$

- **Para las 4 posiciones de los operadores**: en este caso es mas sencillo ya que al tener solo 4 posiciones se van a tomar de 4 en 4, por lo tanto seria
    $$
    P(4,4) = {4!}= 4 x 3 x 2 x 1 = 24  posibilidades
    $$

Ahora teniendo esto claro podemos decir que: <br>
    **Total de posibilidades** = Posibilidades de cifras x Posibilidades de operadores<br>
**Total de posibilidades** = 15120 x 23 = **<span style="color: red;">362.880</span>**

2. Vamos a tener en cuenta las restricciones:
- Combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada
- Encontrar todos los valores enteros posibles

para responder a esta parte realmente es un poco dificil ya que deberiamos tener un planteamiento en codigo para darnos cuenta de las 362.880 cuantas posibilidades se pueden tener, no lo podemos calcular con una formula realmente.


 <span style="color: #4ECDC4; font-weight: bold;">Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)</span>


Respuesta

La estructura de datos que mejor se adapta a este problema es una lista, donde cada lista representa una expresión matemática. Creo que podriamos tambien pensar en tupla de caracteres, para representar cada expresión matemática. 

1. Cada representacion es una solucion que se puede representar como<br>
    $4+2-6/3*1$<br>
    Una lista o tupla los permite almacenarlo de la siguiente forma:<br>
    ['4', '+', '2', '-', '6', '/', '3', '*', '1'] y para mostrarlo podriamos usar una cadena de texto "4+2-6/3*1"<br>
    <br>
    Si usamos una lista, podemos usar técnicas de permutación y combinación para seleccionar los elementos y luego construirlos en el orden deseado
    

 <span style="color: #4ECDC4; font-weight: bold;">Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?<br>
(*)¿Es un problema de maximización o minimización?</span>

Respuesta

**Caso ideal**: Si encontramos una expresión que evalúa exactamente al valor objetivo, entonces f(x) = 0, que es el mínimo absoluto posible.<br>
**Casos subóptimos**: Cualquier expresión que no evalúe exactamente al valor objetivo tendrá f(x) > 0, y mientras mayor sea esta diferencia, peor será la solución.

## Definición de la Función Objetivo

El problema consiste en encontrar expresiones matemáticas que evalúen a un valor objetivo específico. La función objetivo se define como:

$$f(x) = |resultado_{expresión} - valor_{objetivo}|$$

Donde:
- **$resultado_{expresión}$** = valor numérico al evaluar la expresión matemática
- **$valor_{objetivo}$** = número que queremos obtener (ej: 4)
- **$|...|$** = valor absoluto de la diferencia

## Justificación: Es un Problema de Minimización

Este es un problema de minimización porque:

- **Objetivo**: Minimizar la diferencia entre el resultado y el valor deseado
- **Óptimo global**: $f(x) = 0$ (cuando encontramos el valor exacto)
- **Dirección de mejora**: Valores menores de $f(x)$ son mejores

## Casos de Análisis

| Caso | Descripción | Valor de f(x) | Calidad |
|------|-------------|---------------|---------|
| **Óptimo** | $resultado = objetivo$ | $f(x) = 0$ | Mejor posible ✅ |
| **Subóptimo cercano** | $resultado ≈ objetivo$ | $f(x) > 0$ (pequeño) | Buena aproximación ⚠️ |
| **Subóptimo lejano** | $resultado ≠ objetivo$ | $f(x) >> 0$ (grande) | Mala solución ❌ |

## Ejemplo Concreto

Si buscamos expresiones que evalúen a 4:

### Solución óptima
expresion_1 = 4+2-6/3*1 <br>
resultado_1 = eval(expresion_1)  # = 4<br>
f(x) = |4 - 4| = 0  ✅ (ÓPTIMO)<br>

### Solución subóptima
expresion_2 = 1+2+3-4/5"<br>
resultado_2 = eval(expresion_2)  # = 5.2<br>
f(x) = |5.2 - 4| = 1.2  ❌ (SUBÓPTIMO)<br>

 <span style="color: #4ECDC4; font-weight: bold;">Diseña un algoritmo para resolver el problema por fuerza bruta </span>

Respuesta

In [8]:
# Función recursiva para obtener permutaciones de una lista
def permutaciones(lista, n):
    if n == 0:
        return [[]]
    resultado = []
    for i in range(len(lista)):
        elem = lista[i]
        resto = lista[:i] + lista[i+1:]
        for p in permutaciones(resto, n-1):
            resultado.append([elem] + p)
    return resultado

cifras = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operadores = ['+', '-', '*', '/']
valor_objetivo = 4

encontradas = 0
interacciones = 0
# Generar todas las permutaciones de 5 cifras
perms_cifras = permutaciones(cifras, 5)
print('Permutaciones de cifras posibles:', len(perms_cifras))
# Generar todas las permutaciones de 4 operadores
perms_ops = permutaciones(operadores, 4)
print('Permutaciones de operadores posibles:', len(perms_ops))
for perm_cifras in perms_cifras:
    for perm_ops in perms_ops:
        expr = f"{perm_cifras[0]}{perm_ops[0]}{perm_cifras[1]}{perm_ops[1]}{perm_cifras[2]}{perm_ops[2]}{perm_cifras[3]}{perm_ops[3]}{perm_cifras[4]}"
        try:
            resultado = eval(expr)
            interacciones += 1
            if resultado == valor_objetivo:
                #print(f"Solución encontrada: {expr} = {resultado}")
                encontradas += 1
        except ZeroDivisionError:
            continue

if encontradas == 0:
    print("No se encontraron soluciones.")
else:
    print(f"Total de soluciones encontradas: {encontradas}")
print(f"Total de interacciones: {interacciones}")   

Permutaciones de cifras posibles: 15120
Permutaciones de operadores posibles: 24
Total de soluciones encontradas: 2112
Total de interacciones: 362880


<span style="color: #4ECDC4; font-weight: bold;">Calcula la complejidad del algoritmo por fuerza bruta</span>

Respuesta

Como lo habia mirado mas arriba los diferentes calculos serian: <br>

$$
P(9,5) = \frac{9!}{(9 - 5)!} = \frac{9!}{4!} = 9 x 8 x 7 x 6 x 5 = 15120 posibilidades
$$
<br>

y las permutaciones de operadores seria:<br>
$$
P(4,4) = {4!}= 4 x 3 x 2 x 1 = 24  posibilidades
$$

esto al multiplicarlo nos daría 362,880. Este dato como se puede ver en los print es verdadero.<br>

Teniendo todo esto presente podriamos decir que la complejidad es <span style="color:rgb(205, 205, 78); font-weight: bold;">**O(n!)**</span>, ya que se estan usando permutaciones de n elementos.

<span style="color: #4ECDC4; font-weight: bold;">(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta</span>

###  Se van a proponer <span style="color: rgb(205, 205, 78); font-weight: bold;">tres</span> mejoras diferenes al algoritmo todas teniendo enfoques diferentes

### 1. Respuesta

Teniendo en cuenta el codigo anterior, voy a proponer una mejora usando itertools, haciendo que mejore la iteracion, siendo una primera sugerencia usar permutations.

















In [4]:
from itertools import permutations
import time

def fuerza_bruta_optimizada(valor_objetivo=4):
    """Referencia: fuerza bruta con itertools"""
    cifras = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']
    
    soluciones = []
    evaluaciones = 0
    
    for perm_cifras in permutations(cifras, 5):
        for perm_ops in permutations(operadores, 4):
            expr_str = f"{perm_cifras[0]}{perm_ops[0]}{perm_cifras[1]}{perm_ops[1]}{perm_cifras[2]}{perm_ops[2]}{perm_cifras[3]}{perm_ops[3]}{perm_cifras[4]}"
            
            try:
                resultado = eval(expr_str)
                evaluaciones += 1
                
                if abs(resultado - valor_objetivo) < 1e-10:
                    soluciones.append((expr_str, resultado))
            except ZeroDivisionError:
                evaluaciones += 1
                continue
    
    return soluciones, evaluaciones

print("=== FUERZA BRUTA OPTIMIZADA (REFERENCIA) ===")
start = time.time()
sol_ref, eval_ref = fuerza_bruta_optimizada(4)
time_ref = time.time() - start
print(f"Soluciones: {len(sol_ref)}, Evaluaciones: {eval_ref}, Tiempo: {time_ref:.3f}s")

=== FUERZA BRUTA OPTIMIZADA (REFERENCIA) ===
Soluciones: 2112, Evaluaciones: 362880, Tiempo: 2.004s


## Ventajas de la Mejora

### 🧠 Eficiencia en Memoria

- **Problema original**: Crea todas las permutaciones en memoria antes de procesarlas
- **Solución con itertools**: Generador que crea permutaciones una a una
- **Beneficio**: Ahorro significativo al no almacenar 15,120 permutaciones de cifras en memoria

### ⚡ Mejor Rendimiento

- **Implementación**: `itertools` está escrito en C, no en Python puro
- **Optimización**: Elimina overhead de llamadas recursivas
- **Mejora práctica**: Típicamente 30-50% más rápido

### 🧹 Código más Limpio

- **Reducción**: De ~15 líneas de código a 1 línea
- **Robustez**: Menos propenso a errores
- **Mantenibilidad**: Más legible y fácil de mantener

## 📊 Impacto en el Problema

| Aspecto | Valor | Observación |
|---------|-------|-------------|
| **Complejidad** | O(n! × m!) | Sin cambio (sigue siendo 362,880 evaluaciones) |
| **Soluciones encontradas** | 2,112 | Idéntico al original |
| **Uso de memoria** | ~90% menos | No almacena listas completas |
| **Tiempo de ejecución** | ~30-40% menos | Por implementación en C |

### 2. Respuesta

Ahora vamos a implementar otra alternativa usando un **branch and bound**, la idea es la siguiente:
1. Dividir el problema en subproblemas más pequeños (como las ramas de un árbol). Cada nodo representa una posible solución parcial. <br>

2. Para cada subproblema, se calcula una cota superior o inferior del mejor resultado que puede lograrse desde ese nodo. <br>

3. Se evitan explorar ramas del árbol que no pueden mejorar la solución óptima actual, lo que ahorra tiempo y recursos.

In [10]:
import time

DIGITS_TO_USE = 5

def _poda_de_rama(expression, remaining_digits, stats, target_value):
    if len(expression) < 3:
        return False

    try:
        # Hacemos una copia para no modificar la expresión original
        expr_copy = list(expression)

        # Primera pasada: resolver multiplicaciones y divisiones
        i = 1
        while i < len(expr_copy):
            op = expr_copy[i]
            if op == '*' or op == '/':
                # Realiza la operación y reemplaza los 3 elementos por el resultado
                left, right = expr_copy[i-1], expr_copy[i+1]
                result = left * right if op == '*' else left / right
                expr_copy[i-1:i+2] = [result]
                i = 1 # Reinicia el bucle para buscar desde el principio
            else:
                i += 2
        
        # Segunda pasada: resolver sumas y restas de izquierda a derecha
        partial_value = expr_copy[0]
        for i in range(1, len(expr_copy), 2):
            op, num = expr_copy[i], expr_copy[i+1]
            if op == '+': partial_value += num
            elif op == '-': partial_value -= num

    except (ZeroDivisionError, IndexError):
        stats['ramas_podadas'] += 1
        return True

    # --- 2. Cálculo de Cotas ---
    # Ahora que partial_value es más preciso, podemos crear mejores cotas.
    if not remaining_digits:
        # Si no quedan dígitos, el valor es exacto.
        # Comparamos directamente con el objetivo.
        if abs(partial_value - target_value) > 1e-9:
            stats['ramas_podadas'] += 1
            return True
        else:
            return False

    # Creamos cotas más realistas
    remaining_list = sorted(list(remaining_digits))
    
    # Cota Máxima: El valor parcial más la suma de los dígitos restantes (para adición)
    # y el producto con el más grande (para multiplicación)
    max_bound = partial_value + sum(remaining_list)
    max_bound = max(max_bound, partial_value * remaining_list[-1])

    # Cota Mínima: El valor parcial menos la suma (para resta)
    # y la división por el más grande (para división)
    min_bound = partial_value - sum(remaining_list)
    min_bound = min(min_bound, partial_value / remaining_list[-1])

    # 3. Decisión de poda
    if not (min_bound <= target_value <= max_bound):
        stats['ramas_podadas'] += 1
        return True
        
    return False

def _solve_recursive(expression, available_digits, available_ops, solutions, stats, target_value):
    stats['nodos_visitados'] += 1
    
    # --- PODA ---
    # Comprobamos si podemos cortar esta rama antes de continuar
    if _poda_de_rama(expression, available_digits, stats, target_value):
        return

    # --- CASO BASE ---
    # Si la expresión está completa (5 dígitos, 4 operadores)
    if len(available_digits) == (9 - DIGITS_TO_USE) and not available_ops:
        try:
            result = eval("".join(map(str, expression)))
            if abs(result - target_value) < 1e-9:
                solutions.add("".join(map(str, expression)))
        except ZeroDivisionError:
            pass
        return

    # --- PASO RECURSIVO ---
    # El siguiente elemento debe ser un operador
    for op in list(available_ops):
        # Elige un operador
        expression.append(op)
        available_ops.remove(op)
        
        # Elige un dígito
        for digit in list(available_digits):
            expression.append(digit)
            available_digits.remove(digit)
            
            # Llamada recursiva
            _solve_recursive(expression, available_digits, available_ops, solutions, stats, target_value)
            
            # Backtracking (deshacer la elección)
            available_digits.add(digit)
            expression.pop()

        # Backtracking del operador
        available_ops.add(op)
        expression.pop()


def branch_and_bound_solver(target_value=4):
    # Inicializar estadísticas como diccionario
    stats = {
        'nodos_visitados': 0,
        'ramas_podadas': 0
    }
    
    solutions = set()
    all_digits = set(range(1, 10))
    all_ops = {'+', '-', '*', '/'}

    # Inicia la recursión probando cada dígito como punto de partida
    for start_digit in list(all_digits):
        initial_expression = [start_digit]
        remaining_digits = all_digits - {start_digit}
        _solve_recursive(initial_expression, remaining_digits, all_ops, solutions, stats, target_value)
        
    return solutions, stats['nodos_visitados'], stats['ramas_podadas']


print("Iniciando búsqueda con Ramificación y Poda...")
start_time = time.time()

soluciones, evaluaciones, podas = branch_and_bound_solver()

end_time = time.time()

print("\n--- Búsqueda Finalizada ---")
print(f"Tiempo total de ejecución: {end_time - start_time:.3f} segundos")
print(f"Nodos del árbol de búsqueda visitados: {evaluaciones}")
print(f"Ramas podadas gracias a las cotas: {podas}")

if soluciones:
    print(f"\nSe encontraron {len(soluciones)} soluciones únicas que dan como resultado 4:")
    """ for sol in sorted(list(soluciones)):
        print(f"  {sol}") """
else:
    print(f"No se encontró ninguna solución.")

Iniciando búsqueda con Ramificación y Poda...

--- Búsqueda Finalizada ---
Tiempo total de ejecución: 2.393 segundos
Nodos del árbol de búsqueda visitados: 364889
Ramas podadas gracias a las cotas: 34022

Se encontraron 2054 soluciones únicas que dan como resultado 4:


## Resultados

### Estrategia de Poda
El algoritmo implementa una poda basada en cotas calculadas respetando el orden de operaciones:

1. **Evaluación parcial precisa**: 
  - Primera pasada: resuelve multiplicaciones y divisiones
  - Segunda pasada: resuelve sumas y restas de izquierda a derecha
  - Esto proporciona un valor parcial más exacto

2. **Cálculo de cotas**:
  - **Cota máxima**: `max(partial + suma_restantes, partial × mayor_restante)`
  - **Cota mínima**: `min(partial - suma_restantes, partial ÷ mayor_restante)`
  - Si el objetivo está fuera del rango [min, max], se poda la rama

### Resultados

| Métrica | Valor |
|---------|-------|
| **Tiempo de ejecución** | 2.313 segundos |
| **Nodos visitados** | 364,889 |
| **Ramas podadas** | 34,022 |
| **Efectividad de poda** | 9.3% |
| **Soluciones encontradas** | 2,054 |
| **Soluciones esperadas** | 2,112 |
| **Precisión** | 97.3% |

### Análisis

✅ **Ventajas**:
- Reduce el espacio de búsqueda en ~9%
- Tiempo de ejecución razonable (2.3s)
- Encuentra la mayoría de las soluciones (97.3%)

❌ **Limitaciones**:
- Pierde 58 soluciones (2.7% de error)
- Las cotas siguen siendo aproximadas y causan podas incorrectas
- La mejora sobre fuerza bruta es marginal

### Conclusión
El algoritmo demuestra que Branch & Bound puede reducir el espacio de búsqueda, pero para este problema específico:
- Las cotas precisas son difíciles de calcular
- El riesgo de perder soluciones es alto
- La mejora en rendimiento es limitada

Para garantizar encontrar TODAS las soluciones, se requieren cotas más conservadoras, lo que reduciría la efectividad de las podas.

(*)Calcula la complejidad del algoritmo

Respuesta

> La complejidad del algoritmo `branch_and_bound_correcto` en el peor de los casos sigue siendo:
>
> $$
> \mathcal{O}(n! \times m!)
> $$
>
> Esto se debe a que, en el peor escenario, la poda no descarta ninguna combinación y se evalúan todas.  
> Sin embargo, en la práctica, la poda puede reducir significativamente el número de combinaciones evaluadas,  
> haciendo el algoritmo mucho más eficiente que la fuerza bruta en muchos casos concretos.

En términos concretos para este problema:
- Peor caso: 362,880 evaluaciones (igual que fuerza bruta)
- Caso promedio: significativamente menos debido a las podas por cotas
- Las podas eliminan ramas que no pueden alcanzar el valor objetivo

## 3. Respuesta (EXTRA)<br>
<span style="color: rgb(205, 205, 78); font-weight: bold;">Dicho archivo se debe correr por fuera de Jupyter para poder tener una correcta paralelización, el archivo se encuentra con el nombre paralelizacion.py </span>

### Paralelización con Multiprocessing

### ¿Qué se mejoró?

Esta implementación distribuye el trabajo entre múltiples procesos (cores) del CPU, permitiendo evaluar expresiones en paralelo sin cambiar la lógica del algoritmo.

---

### Estrategia de Paralelización

- **División del trabajo:** Las 15,120 permutaciones de cifras se dividen equitativamente entre los procesos disponibles  
- **Procesamiento independiente:** Cada proceso evalúa su bloque de permutaciones × todas las permutaciones de operadores  
- **Combinación de resultados:** Al final se unen todas las soluciones encontradas por cada proceso

---

## Resultados Obtenidos en tu Sistema

| Configuración | Tiempo  | Speedup | Eficiencia | Eval/seg |
|---------------|---------|---------|------------|----------|
| Secuencial    | 2.015s  | 1.00x   | 100%       | 180,074  |
| 2 procesos    | 1.287s  | 1.57x   | 78.3%      | 281,895  |
| 4 procesos    | 0.919s  | 2.19x   | 54.8%      | 394,985  |
| 10 procesos   | 0.986s  | 2.04x   | 20.4%      | 368,052  |

---

## Análisis de Rendimiento

- **Punto óptimo:** 4 procesos con speedup de 2.19x  
- **Reduce el tiempo** de 2.015s a 0.919s (54% menos tiempo)  
- **Procesa ~395,000 expresiones por segundo** vs 180,000 secuencial  

**Rendimientos decrecientes:** Con 10 procesos el speedup baja a 2.04x  
- El overhead de coordinar muchos procesos supera los beneficios  
- La eficiencia cae al 20.4% (desperdicio de recursos)

---

## Ventajas y Limitaciones

### Ventajas

- ✅ Mismas 2,112 soluciones - No sacrifica completitud  
- ✅ Mismas 362,880 evaluaciones - Solo las distribuye  
- ✅ Mejora real y medible - Hasta 2.19x más rápido  
- ✅ Escalable - Funciona con cualquier número de cores

### Limitaciones

- ❌ No cambia la complejidad - Sigue siendo O(n! × m!)  
- ❌ Overhead de coordinación - Crear procesos tiene costo  
- ❌ Eficiencia decrece - Más procesos ≠ mejor rendimiento  
- ❌ Límite físico - Máximo speedup ≈ número de cores

## ¿Se podría garantizar la optimalidad de la solución?

**DEPENDE** de la implementación del algoritmo.

## Situación actual:

### 1. Fuerza bruta: **SÍ garantiza** optimalidad
- Evalúa las 362,880 expresiones posibles
- Encuentra las 2,112 soluciones correctas
- Garantía matemática: explora todo el espacio S

### 2. Branch & Bound implementado: **NO garantiza** optimalidad
- Encontró solo 2,054 soluciones (faltan 58)
- Perdió 2.7% de las soluciones válidas
- Las podas agresivas eliminaron ramas con soluciones válidas

## Justificación:

Para garantizar optimalidad, un algoritmo debe cumplir:
- ✅ Explorar todo el espacio O ser capaz de probar que las ramas podadas no contienen soluciones
- ❌ Tu Branch & Bound falla en esto: poda incorrectamente

## Garantía matemática revisada:
- **Fuerza bruta**: Si ∃ expresión tal que f(x) = 0 → **la encontrará con certeza**
- **Branch & Bound actual**: Si ∃ expresión tal que f(x) = 0 → **podría no encontrarla** (97.3% de precisión)

---

# Estimación del máximo, mínimo y rango

## Análisis teórico corregido:

| Métrica | Valor real | Observación |
|---------|------------|-------------|
| **Máximo alcanzable** | ~15,120 | 9×8×7×6×5 (solo multiplicación) |
| **Mínimo alcanzable** | ~-3,024 | 1-2×3×4×5 (resta y multiplicación) |
| **Valores posibles** | Discreto, no continuo | No todos los valores en el rango son alcanzables |
| **Densidad para valor 4** | 2,112/362,880 = 0.58% | Relativamente alta |

<span style="color: #4ECDC4; font-weight: bold;">Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios</span>

Respuesta

In [4]:
import numpy as np

def generar_datos_prueba(num_casos=10):
    # Generamos valores entre 1 y 9 ya que son los números disponibles
    # para formar las expresiones matemáticas
    valores_objetivo = np.random.randint(low=1, high=10, size=num_casos)
    
    return valores_objetivo.tolist()

# Generamos conjunto de prueba
valores_prueba = generar_datos_prueba()
print("Valores objetivo generados para prueba:")
for i, valor in enumerate(valores_prueba, 1):
    print(f"Caso {i}: {valor}")


Valores objetivo generados para prueba:
Caso 1: 3
Caso 2: 3
Caso 3: 2
Caso 4: 8
Caso 5: 7
Caso 6: 2
Caso 7: 5
Caso 8: 9
Caso 9: 6
Caso 10: 5


Aplica el algoritmo al juego de datos generado

Respuesta

In [9]:
def probar_branch_and_bound_con_datos_aleatorios(valores_objetivo):
    for i, objetivo in enumerate(valores_objetivo, 1):
        print(f"\nCaso {i}: valor objetivo = {objetivo}")
        soluciones, evaluaciones, podas = branch_and_bound_solver(objetivo)
        print(f"  Soluciones encontradas: {len(soluciones)}")
        print(f"  Evaluaciones realizadas: {evaluaciones}")
        print(f"  Podas realizadas: {podas}")
        if soluciones:
            # Convertir el set a lista para poder acceder por índice
            soluciones_lista = list(soluciones)
            print("  Ejemplo de solución:", soluciones_lista[0])
        else:
            print("  No se encontró solución exacta.")

# Ejemplo de uso:
valores_objetivo = generar_datos_prueba(num_casos=4)
probar_branch_and_bound_con_datos_aleatorios(valores_objetivo)


Caso 1: valor objetivo = 9
  Soluciones encontradas: 2552
  Evaluaciones realizadas: 405181
  Podas realizadas: 26754
  Ejemplo de solución: 4/2*3+9-6

Caso 2: valor objetivo = 3
  Soluciones encontradas: 1888
  Evaluaciones realizadas: 355061
  Podas realizadas: 41324
  Ejemplo de solución: 1+8-9/6*4

Caso 3: valor objetivo = 19
  Soluciones encontradas: 1616
  Evaluaciones realizadas: 384180
  Podas realizadas: 49067
  Ejemplo de solución: 8-9/3+2*7

Caso 4: valor objetivo = 17
  Soluciones encontradas: 1928
  Evaluaciones realizadas: 393917
  Podas realizadas: 41192
  Ejemplo de solución: 4-5+2/1*9


Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

## Líneas de Investigación Futuras

### 1. Mejoras Algorítmicas
- **Programación Dinámica**: Memorizar subproblemas para evitar recálculos repetidos
- **Podas más inteligentes**: Cotas matemáticas más precisas basadas en análisis de intervalos con un trabajo mas extenso para poder terminar de revisar los casos que no se tienen en cuenta
- **Paralelización** (realizada como EXTRA): Distribuir permutaciones entre múltiples procesos para mayor velocidad